### DISCUSSION TITLE
AWS ECR login support for OCI helm charts

### DISCUSSION THREAD

**Discussion by leynebe** [argoproj/argo-cd#5248] [Q&A]
Hi all

I'm pretty new to Argo so I might have missed something, but I'm stumbling upon something that I don't think is implemented yet.

AWS recently added support to push Helm charts to ECR. (https://docs.aws.amazon.com/AmazonECR/latest/userguide/push-oci-artifact.html) The first step in it is to enable the helm experimental option, this shows me this helm OCI stuff is pretty new. (https://helm.sh/docs/topics/registries/)

We pushed a generic helm chart to ECR to be used by multiple applications deployed to K8S using ArgoCD.

To accomplish this we wanted to add AWS ECR as a repository source in ArgoCD, that's where errors occurred.

![image](https://user-images.githubusercontent.com/34703148/104581248-96b08f00-565e-11eb-90b4-9a54c32d5a85.png)

I also wasn't able to add the ECR as a repository in the UI, but I was able to add it via the CLI. I cannot test if it works though.

![image](https://user-images.githubusercontent.com/34703148/104586963-80a6cc80-5666-11eb-9ada-3d0f988a1281.png)

There is an additional issue in that "aws ecr get-login-password" only provides the password for 12 hrs. (Which is probably not a problem you should/can solve.)

We're looking at just putting the chart in a git repo and be done with it instead of using AWS ECR.

Thanks for any helmful feedback :)

Ben

**Comment by jannfis** [comment]
Hey @leynebe, it seems you are missing out the protocol on the repository URL. 

Please try to add the repository using a full-qualified URL, starting with `https://` and see if that brings you further forward. Also, judging from your screenshot, you haven't added the repository, but a credential template.

**Reply by leynebe** [reply]
@jannfis thx for the quick reply

I managed to get a step further using your help.

I added the credentials template to take care of auth regarding the ecr repo.
![image](https://user-images.githubusercontent.com/34703148/104594562-9ec5fa00-5671-11eb-8803-a6e880f97756.png)

Then create the project in the UI, but stuck at 404 errors this time:
![image](https://user-images.githubusercontent.com/34703148/104594729-dd5bb480-5671-11eb-861f-313c803a8ab4.png)

And to show that this repository does indeed exist I tried to pull it again:
![image](https://user-images.githubusercontent.com/34703148/104595025-480cf000-5672-11eb-8a00-c5cbabf3be1f.png)

Any more recommendations?

**Comment by jannfis** [comment]
I think you will need to add the Helm repository explicitly to the Argo CD configuration and enable OCI support for it.

Since this feature is not (yet) exposed in the web UI, you will have to do it using the CLI's `--enable-oci` option, e.g.

```bash
argocd repo add --type helm --name <repo-name> --enable-oci <your-repository-url>
```

If you get authentication failure, try to pass `--username` and `--password` explicitly in above example.

**Reply by mlopez-eb** [reply]
username/password doesn't help in this situation because it changes every 12 hours

**Reply by TomKeur** [reply]
Just saw this, but in my case Argo CD is not running on AWS, so I'm using External Secrets Operator, with "ECRAuthorizationToken", this is auto refreshing the token every 12H when it gets expired.

I've followed this example (but in total, but it should give you a good direction).
https://medium.com/@prassonmishra330/helm-chart-deployment-with-argocd-and-ecr-f8f95f625a56

It's this resource in External Secrets Operator:
https://external-secrets.io/v0.19.2/api/generator/ecr/
